src/audio/kpb.c

```diff
@@ -527,17 +536,105 @@ static void kpb_init_draining(struct comp_data *kpb, struct kpb_client *cli)
 	} else if (!is_sink_ready) {
 		trace_kpb_error("kpb_init_draining() error: "
 				"sink not ready for draining");
+
+	} else if (!kpb_have_enough_history_data(buff, history_depth)) {
+		trace_kpb_error("kpb_init_draining() error: "
+				"not enough data in history buffer");
 	} else {
+		/* draining accepted, find proper buffer to start reading */
+		while (buff->next && buff->next != first_buff) {
+			if (buff->state == KPB_BUFFER_FREE)
+				break;
+			else
+				buff = buff->next;
+		}
+
+		while (buffered < history_depth) {
+			if (buff->state == KPB_BUFFER_FULL) {
+				buffered += (int)buff->end_addr -
```

Comment:
casting to int means any buffer above 0x80000000 in memory will be negative. end_addr >= start_address anyway.